feat: Exchange JWT token with session cookies

jawad-khan opened the pull request:
Exchange JWT token with session cookies

## Description

Exchange JWT token with session cookies so that it can work on mobile.
Mobile platform is moving to JWT and to authenticate xblocks we need session cookies
in exchange of JWT token.

[LEARNER-8518](https://openedx.atlassian.net/browse/LEARNER-8518)

## Review pre-merge checks

Additional details in the[ Mobile JWT Confluence page](https://openedx.atlassian.net/wiki/spaces/AC/pages/3351609433/Mobile+authentication+with+JWTs).

For additional details, see:
https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-mobile-migration-to-jwt.rst

- [x] Requires asymmetrically signed JWT.
- [x] Grant_type added to JWT and checked.
- [x] JWT related calls using library functions.
- [x] Ensure user account is not disabled.
- [x] Ensure JWT expiration reduced to one hour (may block usage, but not merging?)
- [x] Ensure ADR exists and is shared

robrap commented on openedx/core/djangoapps/auth_exchange/README.rst:
```diff
@@ -10,4 +10,4 @@ The following are currently implemented:
 2. LoginWithAccessTokenView
 
-   1st party (open-edx) OAuth 2.0 access token -> session cookie
+   1st party (open-edx) OAuth 2.0 access/jwt token -> session cookie
```
Is this more accurate?
```suggestion
   1st party (open-edx) OAuth 2.0 access token (bearer/jwt) -> session cookie
```

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -150,22 +151,27 @@ def setUp(self):
 
-    def _verify_response(self, access_token, expected_status_code, expected_cookie_name=None):
+    def _verify_response(self, access_token, expected_status_code,
+                         token_type='Bearer', expected_cookie_name=None, data=None):
```
Nit: if this argument order makes the most sense to you, having callers use this consistent order would be nice. Lots of callers use `data` before `token_type`, as an example.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
-        if not self._is_grant_password(request.auth):
+        access_token = request.auth
+        is_jwt_authenticated = isinstance(request.successful_authenticator, JwtAuthentication)
+        if is_jwt_authenticated:
```
1. I'm not sure this is enough. JwtAuthentication allows both header and cookie auth, and I _think_ cookie auth is also a fall back? So, someone could supply a JWT token in the header that gets the right access token from DB, but isn't actually authenticated? I may need to talk to @timmc-edx about this.
2. Nit: I find the following easier to read, using the new [walrus operator](https://realpython.com/python-walrus-operator/).
```suggestion
        if is_jwt_authenticated := isinstance(request.successful_authenticator, JwtAuthentication):
```

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
             request.user.backend = self._get_path_of_arbitrary_backend_for_user(request.user)
 
-        if not self._is_grant_password(request.auth):
+        access_token = request.auth
```
Maybe this could move to an `else` case below, with a comment about it being a bearer token?

jawad-khan replied:
Sure.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
+        access_token = request.auth
+        is_jwt_authenticated = isinstance(request.successful_authenticator, JwtAuthentication)
+        if is_jwt_authenticated:
+            refresh_token = dot_models.RefreshToken.objects.filter(token=request.data.get('refresh_token')).first()
```
Can you explain why and how you think this refresh_token necessarily relates to the current JWT token? Or enough to do the later check? Does it matter if it isn't the right one? Seems like it would.

robrap replied:
Better yet, is the access token associated to the refresh token in the DB unused in the case of a JWT access token request? Can we have DOT instead store the real JWT access token? If it did, could this code work unchanged?

robrap replied:
Searches lead me to: https://github.com/jazzband/django-oauth-toolkit/pull/467

jawad-khan replied:
According to the current user flow this refresh token relates to the access_token. We can have a discussion on this.

robrap replied:
* I think it will be safer and cleaner to add the grant type to the JWT itself and be able to check it directly. 
* I updated the [Mobile JWT Confluence doc](https://openedx.atlassian.net/wiki/spaces/AC/pages/3351609433/Mobile+authentication+with+JWTs) with this detail, and other notes about changes that are required.
* I added a checklist to the PR description of updates we want to make.

robrap replied:
Marking resolved.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
+        is_jwt_authenticated = isinstance(request.successful_authenticator, JwtAuthentication)
+        if is_jwt_authenticated:
+            refresh_token = dot_models.RefreshToken.objects.filter(token=request.data.get('refresh_token')).first()
+            if refresh_token:
```
This would need an `else` case. Should it raise an exception? Other?

robrap reviewed (commented):
Initial review. Thanks.

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -180,3 +184,47 @@ def test_dot_password_grant_supported(self):
+            jwt_token = self._create_jwt_token()
+
+        # wait for jwt to expire before sending the request
+        sleep(6)
```
Sleeping in tests is a bad practice because we want tests to be fast. Search the tests for `freezegun` and see if you can find a better way to adjust the times appropriately.

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -180,3 +184,47 @@ def test_dot_password_grant_supported(self):
         access_token = self._create_dot_access_token()
         self._verify_response(access_token, expected_status_code=401)
+
+    def _create_jwt_token(self, grant_type=Application.GRANT_PASSWORD, scope='email profile'):
```
Can you please change the following, at least in the test code.
```suggestion
    def _create_jwt_token(self, grant_type='password', scope='email profile'):
```

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
             request.user.backend = self._get_path_of_arbitrary_backend_for_user(request.user)
 
-        if not self._is_grant_password(request.auth):
+        if isinstance(request.successful_authenticator, JwtAuthentication):
```
There is a library call for this as well.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
+            is_grant_password = jwt_payload['grant_type'] == dot_models.Application.GRANT_PASSWORD
+        else:
+            # If not jwt then it must be bearer token
+            is_grant_password = self._is_grant_password(request.auth)
```
Would it be cleaner for `_is_grant_password()` to be updated to take a request, determine if it is JWT or Bearer and return the correct result?

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,7 +146,14 @@ def post(self, request):
+        if not is_grant_password:
             raise AuthenticationFailed({
                 'error_code': 'non_supported_token',
                 'developer_message': 'Only support DOT type access token with grant type password. '
```
```suggestion
                'developer_message': 'Only access tokens with grant type password are supported. '
```

robrap reviewed (commented):
This is at least a pass on the review. I may have missed things. It would be simpler if this branch could be based off the other branch so those changes don't appear, but that takes a little git knowledge.

jawad-khan committed:
feat: Exchange jwt token with session cookies

Exchange jwt token with session cookies so that it can work on mobile.
Mobile platform is migrating to jwt and for accessing  xblocks we need session cookies in exchnage of jwt token.
For additional details, see
https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-mobile-migration-to-jwt.rst

LEARNER-8518

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -120,14 +125,20 @@ def _get_path_of_arbitrary_backend_for_user(user):
+        if is_jwt_authenticated(request):
+            jwt_payload = get_decoded_jwt_from_auth(request)
+            grant_type = jwt_payload['grant_type']
+            if grant_type and grant_type == dot_models.Application.GRANT_PASSWORD:
```
Nit: could his just be:
```
if jwt_payload['grant_type'] == dot_models.Application.GRANT_PASSWORD:
```

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -180,3 +185,53 @@ def test_dot_password_grant_supported(self):
+
+        assert 'session_key' not in self.client.session
+
+    def test_valid_jwt_with_password_grant(self):
```
* Consider renaming test to `test_success_with_valid_jwt_with_password_grant` and other tests with prefix `test_failure_with...", like `test_failure_with_valid_symmetric_jwt`.
* Consider moving this test to the end so all the failure cases are together.

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -180,3 +185,53 @@ def test_dot_password_grant_supported(self):
+        self._verify_response(jwt_token, token_type="JWT",
+                              expected_status_code=204, expected_cookie_name='sessionid')
+
+        assert int(self.client.session['_auth_user_id']) == self.user.id
```
Maybe also add `assert 'session_key' in self.client.session` to show the opposite assertion of the failure assertions?

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -145,12 +158,24 @@ def post(self, request):
+                    'error_code': 'non_asymmetric_token',
+                    'developer_message': 'Only asymmetric jwt are supported.'
+                }
+                raise AuthenticationFailed(error_msg) from symmetric_jwt_error
```
Do you want the test to confirm that you are actually getting this error, and not some other unexpected error?

robrap commented on openedx/core/djangoapps/oauth_dispatch/views.py:
```diff
@@ -117,7 +117,9 @@ def _get_jwt_content_from_access_token_content(self, request, response):
         """
         opaque_token_dict = json.loads(response.content.decode('utf-8'))
-        jwt_token_dict = create_jwt_token_dict(opaque_token_dict, self.get_adapter(request))
+        use_asymmetric_key = request.POST.get('asymmetric_jwt', False)
+        jwt_token_dict = create_jwt_token_dict(opaque_token_dict, self.get_adapter(request),
```
Can we get test coverage for the new parameter?

robrap reviewed (commented):
Looks great. Just some test related comments.

robrap commented:
@jawad-khan: In addition to the above comments, even though it was checked off in the PR description, are we missing "Ensure user account is not disabled."? Note: I unchecked it for now.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -128,21 +128,45 @@ def _get_path_of_arbitrary_backend_for_user(user):
+    def _raise_exception_if_grant_is_not_password(request):
         """
         Check if the access token provided is DOT based and has password type grant.
         """
```
1. Nit: Possible function rename to be positive instead of negative.
2. Updated docstring
```suggestion
    def _ensure_access_token_has_password_grant(request):
        """
        Ensures the access token provided has password type grant.
        """
```

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -128,21 +128,45 @@ def _get_path_of_arbitrary_backend_for_user(user):
-        return False
+        raise AuthenticationFailed({
+            'error_code': 'non_supported_token',
+            'developer_message': 'Only access tokens with grant type password are supported. '
```
Nit: removed space at end of string.
```suggestion
            'developer_message': 'Only access tokens with grant type password are supported.'
```

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -128,21 +128,45 @@ def _get_path_of_arbitrary_backend_for_user(user):
+        })
+
+    @staticmethod
+    def _raise_exception_if_jwt_is_symmetric(request):
```
Nit: What do you feel about this suggestion?
```suggestion
    def _ensure_jwt_is_asymmetric(request):
```

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -128,21 +128,45 @@ def _get_path_of_arbitrary_backend_for_user(user):
+            error_msg = {
+                'error_code': 'account_disabled',
+                'developer_message': 'user account is disabled'
+            }
```
Consider adding a `log.info` with something like 'Login exchange failed because user account XXX was disabled.'
1. If needed, replace "Login exchange" in my message with something more accurate.
2. This would provide a safe way to capture the user id if we needed it for security purposes.
3. You do not need to add this to the test.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -128,21 +128,45 @@ def _get_path_of_arbitrary_backend_for_user(user):
+                raise AuthenticationFailed(error_msg) from symmetric_jwt_error
+
+    @staticmethod
+    def _raise_exception_if_user_is_disabled(request):
```
Nit: What do you feel about this suggestion? In this case, I think it more clear with `not` than using `enabled`.
```suggestion
    def _ensure_user_is_not_disabled(request):
```

robrap commented on openedx/core/djangoapps/auth_exchange/tests/test_views.py:
```diff
@@ -167,16 +175,85 @@ def _create_dot_access_token(self, grant_type='Client credentials'):
+
+        assert '_auth_user_id' not in self.client.session
+
+    def test_failure_with_valid_jwt_with_not_password_grant(self):
```
Nit:
```suggestion
    def test_failure_with_valid_jwt_but_not_password_grant(self):
```

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -337,10 +343,17 @@ def test_jwt_access_token_scopes_and_filters(self, grant_type):
 
+    def test_asymmetric_jwt_access_token(self):
+        """
+        Verify the JWT is asymmetric.
```
Nit:
```suggestion
        Verify the JWT is asymmetric when requested.
```
Hopefully it is easy for you to just accept these changes from inside github. I know they are very minor nits, which you can also ignore.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -133,7 +134,7 @@ def setUp(self):
         self.view_class = views.AccessTokenView
 
-    def _post_body(self, user, client, token_type=None, scope=None):
+    def _post_body(self, user, client, token_type=None, scope=None, asymmetric_jwt=False):
```
Would `asymmetric_jwt=None` better test the backward compatible case, where it wasn't requested? I think the three cases are (not supplied, supplied as True, and supplied as False) and I do not think we need coverage for the third case, but I do think we want it for the first case, which is the general case.

robrap commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -350,14 +363,17 @@ def setUp(self):
         super().setUp()
 
-    def _post_body(self, user, client, token_type=None, scope=None):
+    def _post_body(self, user, client, token_type=None, scope=None, asymmetric_jwt=False):
```
I'm not clear on when this gets used, and if it is needed at all, and if it should also be `None`. See my other related comment.

robrap reviewed (commented):
Nice. Thank you. Lots of minor nits, but very close.

robrap commented on openedx/core/djangoapps/auth_exchange/views.py:
```diff
@@ -160,11 +167,15 @@ def _raise_exception_if_jwt_is_symmetric(request):
-    def _raise_exception_if_user_is_disabled(request):
+    def _ensure_user_is_not_disabled(request):
+        """
+        Ensures the user requesting for session cookies is not disabled..
```
Minor typo.
```suggestion
        Ensures the user requesting for session cookies is not disabled.
```

robrap reviewed (approved):
Great job. One minor fix. Then feel free to squash. Since we don't have a toggle, please merge when you are ready to watch and rollback if we cause any auth issues.

robrap commented:
@jawad-khan: Maybe add a link to the ADR for more details in the PR description and final squashed commit comment.
